File: java/0017-letter-combinations-of-phone-number.java
import java.util.*;

/**
 * LeetCode Problem 0017: Letter Combinations of a Phone Number
 *
 * Problem:
 * Given a string containing digits from 2-9 inclusive, return all possible letter
 * combinations that the number could represent.
 *
 * Digit mapping:
 * 2 -> "abc", 3 -> "def", 4 -> "ghi", 5 -> "jkl",
 * 6 -> "mno", 7 -> "pqrs", 8 -> "tuv", 9 -> "wxyz"
 *
 * Examples:
 * 1. Input: digits = "23", Output: ["ad","ae","af","bd","be","bf","cd","ce","cf"]
 * 2. Input: digits = "", Output: []
 */
public class LetterCombinationsPhoneNumber {

    /**
     * Q1: Recursive backtracking
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsBacktrack(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        backtrack(digits, 0, "", result, mapping);
        return result;
    }

    private static void backtrack(String digits, int index, String current,
                                  List<String> result, String[] mapping) {
        if (index == digits.length()) {
            result.add(current);
            return;
        }

        String letters = mapping[digits.charAt(index) - '0'];
        for (char letter : letters.toCharArray()) {
            backtrack(digits, index + 1, current + letter, result, mapping);
        }
    }

    /**
     * Q2: Iterative approach
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsIterative(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        result.add("");

        for (char digit : digits.toCharArray()) {
            List<String> temp = new ArrayList<>();
            String letters = mapping[digit - '0'];

            for (String combo : result) {
                for (char letter : letters.toCharArray()) {
                    temp.add(combo + letter);
                }
            }

            result = temp;
        }

        return result;
    }

    /**
     * Q3: BFS with queue
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsBFS(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        Queue<String> queue = new LinkedList<>();
        queue.offer("");

        for (char digit : digits.toCharArray()) {
            int size = queue.size();
            String letters = mapping[digit - '0'];

            for (int i = 0; i < size; i++) {
                String combo = queue.poll();
                for (char letter : letters.toCharArray()) {
                    queue.offer(combo + letter);
                }
            }
        }

        result.addAll(queue);
        return result;
    }

    /**
     * Q4: Using HashMap
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsHashMap(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        Map<Character, String> map = new HashMap<>();
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");

        result.add("");

        for (char digit : digits.toCharArray()) {
            List<String> temp = new ArrayList<>();
            String letters = map.get(digit);

            for (String combo : result) {
                for (char letter : letters.toCharArray()) {
                    temp.add(combo + letter);
                }
            }

            result = temp;
        }

        return result;
    }

    /**
     * Q5: DFS with helper function
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsDFS(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        dfs(digits, 0, "", result, mapping);
        return result;
    }

    private static void dfs(String digits, int index, String current,
                            List<String> result, String[] mapping) {
        if (index == digits.length()) {
            result.add(current);
            return;
        }

        String letters = mapping[digits.charAt(index) - '0'];
        for (char letter : letters.toCharArray()) {
            dfs(digits, index + 1, current + letter, result, mapping);
        }
    }

    /**
     * Q6: StringBuilder approach
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsStringBuilder(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        helper(digits, 0, new StringBuilder(), result, mapping);
        return result;
    }

    private static void helper(String digits, int index, StringBuilder current,
                               List<String> result, String[] mapping) {
        if (index == digits.length()) {
            result.add(current.toString());
            return;
        }

        String letters = mapping[digits.charAt(index) - '0'];
        for (char letter : letters.toCharArray()) {
            current.append(letter);
            helper(digits, index + 1, current, result, mapping);
            current.deleteCharAt(current.length() - 1);
        }
    }

    /**
     * Q7: Stream-based approach
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsStream(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        result.add("");

        for (char digit : digits.toCharArray()) {
            String letters = mapping[digit - '0'];
            result = result.stream()
                    .flatMap(combo -> letters.chars()
                            .mapToObj(c -> combo + (char) c))
                    .collect(java.util.stream.Collectors.toList());
        }

        return result;
    }

    /**
     * Q8: Array-based mapping
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsArray(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        backtrackArray(digits, 0, new char[digits.length()], result, mapping);
        return result;
    }

    private static void backtrackArray(String digits, int index, char[] current,
                                       List<String> result, String[] mapping) {
        if (index == digits.length()) {
            result.add(new String(current));
            return;
        }

        String letters = mapping[digits.charAt(index) - '0'];
        for (char letter : letters.toCharArray()) {
            current[index] = letter;
            backtrackArray(digits, index + 1, current, result, mapping);
        }
    }

    /**
     * Q9: Loop with list manipulation
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsLoop(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        result.add("");

        for (int i = 0; i < digits.length(); i++) {
            String letters = mapping[digits.charAt(i) - '0'];
            int size = result.size();

            for (int j = 0; j < size; j++) {
                String combo = result.remove(0);
                for (char letter : letters.toCharArray()) {
                    result.add(combo + letter);
                }
            }
        }

        return result;
    }

    /**
     * Q10: Optimized with capacity pre-allocation
     *
     * Time Complexity: O(4^n)
     * Space Complexity: O(4^n)
     */
    public static List<String> letterCombinationsOptimized(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;

        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

        int capacity = 1;
        for (char digit : digits.toCharArray()) {
            capacity *= mapping[digit - '0'].length();
        }

        result.add("");

        for (char digit : digits.toCharArray()) {
            String letters = mapping[digit - '0'];
            int size = result.size();
            result.ensureCapacity(size * letters.length());

            for (int i = 0; i < size; i++) {
                String combo = result.get(i);
                for (int j = 1; j < letters.length(); j++) {
                    result.add(combo + letters.charAt(j));
                }
                result.set(i, combo + letters.charAt(0));
            }
        }

        return result;
    }

    // Test cases and main function
    public static void main(String[] args) {
        String[] testCases = {"23", "", "2", "234", "42"};
        int[] expectedSizes = {9, 0, 3, 27, 9};

        java.lang.reflect.Method[] methods = LetterCombinationsPhoneNumber.class.getDeclaredMethods();
        List<java.lang.reflect.Method> testMethods = new ArrayList<>();

        for (java.lang.reflect.Method method : methods) {
            if (method.getName().startsWith("letterCombinations") &&
                method.getReturnType() == List.class) {
                testMethods.add(method);
            }
        }

        Collections.sort(testMethods, Comparator.comparing(java.lang.reflect.Method::getName));

        int questionNum = 1;
        for (java.lang.reflect.Method method : testMethods) {
            System.out.println("\nQ" + questionNum + ": " + method.getName());
            System.out.println("-".repeat(70));

            boolean allPassed = true;
            for (int i = 0; i < testCases.length; i++) {
                try {
                    List<String> result = (List<String>) method.invoke(null, testCases[i]);
                    boolean passed = result.size() == expectedSizes[i];
                    allPassed = allPassed && passed;
                    String status = passed ? "✓" : "✗";
                    String display = result.size() <= 5 ? result.toString() :
                            "[" + result.get(0) + ", " + result.get(1) + ", ..., " + result.get(result.size() - 1) + "]";
                    System.out.println("  " + status + " letterCombinations(\"" + testCases[i] +
                        "\") = " + display + " (size: " + result.size() + ")");
                } catch (Exception e) {
                    System.out.println("  ✗ Error: " + e.getMessage());
                    allPassed = false;
                }
            }

            if (allPassed) {
                System.out.println("  All tests passed!");
            }

            questionNum++;
        }
    }
}

File: python/0017-letter-combinations-of-phone-number.py
"""
LeetCode Problem 0017: Letter Combinations of a Phone Number

Problem:
Given a string containing digits from 2-9 inclusive, return all possible letter
combinations that the number could represent. Return the answer in any order.

A mapping of digit to letters (just like on the telephone buttons):
- 2 -> "abc"
- 3 -> "def"
- 4 -> "ghi"
- 5 -> "jkl"
- 6 -> "mno"
- 7 -> "pqrs"
- 8 -> "tuv"
- 9 -> "wxyz"

Examples:
1. Input: digits = "23", Output: ["ad","ae","af","bd","be","bf","cd","ce","cf"]
2. Input: digits = "", Output: []
3. Input: digits = "2", Output: ["a","b","c"]
4. Input: digits = "234", Output: ["adg","adh","adi","aeg","aeh","aei","afg","afh","afi","bdg","bdh","bdi","beg","beh","bei","bfg","bfh","bfi","cdg","cdh","cdi","ceg","ceh","cei","cfg","cfh","cfi"]

Constraints:
- 0 <= digits.length <= 4
- digits[i].isdigit() and digits[i] != '0' and digits[i] != '1'
"""

# Q1: Recursive backtracking
def letterCombinations_backtrack(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n) where n is length of digits
    Space Complexity: O(4^n) for result storage

    Approach: Use recursion to generate all combinations.
    """
    if not digits:
        return []

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    result = []

    def backtrack(index, current):
        if index == len(digits):
            result.append(current)
            return

        letters = mapping[digits[index]]
        for letter in letters:
            backtrack(index + 1, current + letter)

    backtrack(0, "")
    return result


# Q2: Iterative approach
def letterCombinations_iterative(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Build combinations iteratively.
    """
    if not digits:
        return []

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    result = ['']

    for digit in digits:
        temp = []
        for combo in result:
            for letter in mapping[digit]:
                temp.append(combo + letter)
        result = temp

    return result


# Q3: BFS with queue
def letterCombinations_bfs(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Use BFS with a queue.
    """
    if not digits:
        return []

    from collections import deque

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    queue = deque([''])

    for digit in digits:
        size = len(queue)
        for _ in range(size):
            combo = queue.popleft()
            for letter in mapping[digit]:
                queue.append(combo + letter)

    return list(queue)


# Q4: Using itertools.product
def letterCombinations_product(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Use itertools.product for Cartesian product.
    """
    if not digits:
        return []

    from itertools import product

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    digit_letters = [mapping[d] for d in digits]
    return [''.join(combo) for combo in product(*digit_letters)]


# Q5: DFS with helper function
def letterCombinations_dfs(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: DFS with explicit helper function.
    """
    if not digits:
        return []

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    def dfs(index):
        if index == len(digits):
            return ['']

        result = []
        for letter in mapping[digits[index]]:
            for combo in dfs(index + 1):
                result.append(letter + combo)

        return result

    return dfs(0)


# Q6: List comprehension
def letterCombinations_listcomp(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Use list comprehension for concise code.
    """
    if not digits:
        return []

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    result = ['']
    for digit in digits:
        result = [combo + letter for combo in result for letter in mapping[digit]]

    return result


# Q7: Recursive with list building
def letterCombinations_recursive_list(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Recursive approach with explicit list building.
    """
    if not digits:
        return []

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    def combine(index):
        if index == len(digits):
            return ['']

        rest = combine(index + 1)
        result = []

        for letter in mapping[digits[index]]:
            for combo in rest:
                result.append(letter + combo)

        return result

    return combine(0)


# Q8: Generator approach
def letterCombinations_generator(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Use generators for lazy evaluation.
    """
    if not digits:
        return []

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    def gen_combinations(index, current):
        if index == len(digits):
            yield current
        else:
            for letter in mapping[digits[index]]:
                yield from gen_combinations(index + 1, current + letter)

    return list(gen_combinations(0, ''))


# Q9: Dictionary mapping with recursion
def letterCombinations_dict_recursive(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Dictionary-based mapping with recursion.
    """
    if not digits:
        return []

    phone_map = {
        '2': ['a', 'b', 'c'],
        '3': ['d', 'e', 'f'],
        '4': ['g', 'h', 'i'],
        '5': ['j', 'k', 'l'],
        '6': ['m', 'n', 'o'],
        '7': ['p', 'q', 'r', 's'],
        '8': ['t', 'u', 'v'],
        '9': ['w', 'x', 'y', 'z']
    }

    def helper(index, path):
        if index == len(digits):
            return [path]

        result = []
        for letter in phone_map[digits[index]]:
            result.extend(helper(index + 1, path + letter))

        return result

    return helper(0, '')


# Q10: Tail recursion approach
def letterCombinations_tail_recursive(digits: str) -> list[str]:
    """
    Time Complexity: O(4^n)
    Space Complexity: O(4^n)

    Approach: Tail recursive approach with accumulator.
    """
    if not digits:
        return []

    mapping = {
        '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
        '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
    }

    def tail_rec(index, accumulator):
        if index == len(digits):
            return accumulator

        new_accumulator = []
        for combo in accumulator:
            for letter in mapping[digits[index]]:
                new_accumulator.append(combo + letter)

        return tail_rec(index + 1, new_accumulator)

    return tail_rec(0, [''])


if __name__ == "__main__":
    test_cases = [
        ("23", ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]),
        ("", []),
        ("2", ["a", "b", "c"]),
        ("234", ["adg", "adh", "adi", "aeg", "aeh", "aei", "afg", "afh", "afi",
                 "bdg", "bdh", "bdi", "beg", "beh", "bei", "bfg", "bfh", "bfi",
                 "cdg", "cdh", "cdi", "ceg", "ceh", "cei", "cfg", "cfh", "cfi"]),
        ("2345", ["adgj", "adgk", "adgl", "adhj", "adhk", "adhl", "adij", "adik", "adil",
                  "aegj", "aegk", "aegl", "aehj", "aehk", "aehl", "aeij", "aeik", "aeil",
                  "afgj", "afgk", "afgl", "afhj", "afhk", "afhl", "afij", "afik", "afil",
                  "bdgj", "bdgk", "bdgl", "bdhj", "bdhk", "bdhl", "bdij", "bdik", "bdil",
                  "begj", "begk", "begl", "behj", "behk", "behl", "beij", "beik", "beil",
                  "bfgj", "bfgk", "bfgl", "bfhj", "bfhk", "bfhl", "bfij", "bfik", "bfil",
                  "cdgj", "cdgk", "cdgl", "cdhj", "cdhk", "cdhl", "cdij", "cdik", "cdil",
                  "cegj", "cegk", "cegl", "cehj", "cehk", "cehl", "ceij", "ceik", "ceil",
                  "cfgj", "cfgk", "cfgl", "cfhj", "cfhk", "cfhl", "cfij", "cfik", "cfil"]),
        ("42", ["ga", "gb", "gc", "ha", "hb", "hc", "ia", "ib", "ic"]),
    ]

    functions = [
        letterCombinations_backtrack,
        letterCombinations_iterative,
        letterCombinations_bfs,
        letterCombinations_product,
        letterCombinations_dfs,
        letterCombinations_listcomp,
        letterCombinations_recursive_list,
        letterCombinations_generator,
        letterCombinations_dict_recursive,
        letterCombinations_tail_recursive,
    ]

    for i, func in enumerate(functions, 1):
        print(f"\nQ{i}: {func.__name__}")
        print("-" * 70)
        all_passed = True
        for digits, expected in test_cases:
            result = func(digits)
            result.sort()
            expected.sort()
            passed = result == expected
            all_passed = all_passed and passed
            status = "✓" if passed else "✗"
            display_result = result if len(result) <= 5 else result[:5] + ["..."]
            print(f"  {status} letterCombinations('{digits}') = {display_result} (length: {len(result)})")

        if all_passed:
            print(f"  All tests passed!")
